File: 740. Delete and Earn/dp.js
/**
 * @param {number[]} nums
 * @return {number}
 */
 var deleteAndEarn = function(nums) {
  
    const freq = new Map();
    for(const num of nums){
        freq.set(num, 1 + (freq.get(num) || 0));
    }

    const set = new Set(nums);
    const unique = [...set.keys()].sort((a, b) => a - b);
    const n = unique.length;
    const dp = [...Array(n+1)].fill(0);
    dp[1] = unique[0] * freq.get(unique[0]);
    
    for(let i = 1; i < n; i++){
        const dpIdx = i + 1;
        
        const lastValid = (unique[i-1] + 1 === unique[i]) ? dpIdx - 2 : dpIdx - 1;
        
        dp[dpIdx] = Math.max(
            dp[dpIdx - 1],
            unique[i] * freq.get(unique[i]) + dp[lastValid]
        )
    }
    
    return dp[n];
};

File: 740. Delete and Earn/dp.py
class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        freq = {}
        for num in nums:
            freq[num] = freq.get(num, 0) + 1;
        
        keys = sorted(freq.keys())
        n = len(keys)
        
        dp = [0 for i in range(n+1)];
                
        for i in range(0, n):
            p = i + 1
            num = keys[i]
            take = num * freq[num]
            if i == 0:
                dp[p] = take
            else:
                dp[p] = max(
                    dp[p-1],
                    take + (dp[p-2] if keys[i] - 1 == keys[i-1] else dp[p-1])
                )
        return dp[n];
            
        